DO NOT MERGE: Spike for separating out platform-specific elements

idg10 opened the pull request:
This PR is purely for discussion. It is not intended to be merged in its current form.

The decision to unify Rx into a single System.Reactive component is now causing problems for UI frameworks that targeting netX.0-windows and which aren't WPF or Windows Forms: both WPF and Window Forms get brought along for the ride, which causes tens of megabytes of bloat with self-contained packaging models.

This shows what would be required to split the UI-framework-specific parts back out again.

There is an ADR to discuss the design. It describes how some of the context that led to the current design has changed. However, it's not yet entirely clear whether all the problems can be dealt with.

idg10 committed:
Spike for separating out platform-specific elements

The decision to unify Rx into a single System.Reactive component is now causing problems for UI frameworks that targeting netX.0-windows and which aren't WPF or Windows Forms: both WPF and Window Forms get brought along for the ride, which causes tens of megabytes of bloat with self-contained packaging models.

idg10 committed:
Fix markdown snafu

HowardvanRooijen commented on Rx.NET/Documentation/adr/0001-multi-platform-packaging.md:
```diff
@@ -0,0 +1,106 @@
+    * `System.Reactive.WindowsRuntime` (in the [`Rx-Metro`](https://www.nuget.org/packages/Rx-Metro/2.2.5) NuGet package)—(conversions between WinRT types and)
+* v3 changed the NuGet package names, but essentially kept the same structure
+  * The old `Rx-Main` metapackage is replaced by `System.Reactive`
+  * `Rx-Interfaces` becomes `System.Reactive.Interfaces` (v3.0, not clear if there were actually interface changes)
```
Are these packages or are these now type forwarding facades?

idg10 replied:
No, the type forwarding starts in v4.

HowardvanRooijen commented on Rx.NET/Documentation/adr/0001-multi-platform-packaging.md:
```diff
@@ -0,0 +1,106 @@
+# NuGet packages and multi-platform support
+
+Rx has tried a few strategies for cross-platform support over its history. The approach used in v5.0 turned out to cause problems for some UI frameworks: self-contained deployments would end up including copies of various WPF and Windows Forms components whether they used them or not, causing binaries to be tends of megabytes larger than they otherwise would have been.
```
Can you add the avalonia issue link here?

HowardvanRooijen commented on Rx.NET/Documentation/adr/0001-multi-platform-packaging.md:
Could you add in brackets (.NET 7.0) so in the future we know what version of .NET was current at the time?

HowardvanRooijen commented on Rx.NET/Source/src/System.Reactive/System.Reactive.csproj:
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="MSBuild.Sdk.Extras">
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp3.1;netstandard2.0;net472;uap10.0.16299;net5.0;net5.0-windows10.0.19041</TargetFrameworks>
+    <TargetFrameworks>net6.0;netstandard2.0;net472;uap10.0.16299</TargetFrameworks>
```
Any idea why it supports 472 rather than 462 (as 462 was the first fully compliant netstandard2.0 version of the .NET Framework)?

HowardvanRooijen commented on Rx.NET/Source/tests/Tests.System.Reactive/Tests.System.Reactive.csproj:
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="MSBuild.Sdk.Extras">
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp3.1;net472;netcoreapp2.1;net5.0;net5.0-windows10.0.19041</TargetFrameworks>
+    <TargetFrameworks>net472;net6.0-windows10.0.19041</TargetFrameworks>
```
do we need a net7.0-windows TFM?

HowardvanRooijen commented on Rx.NET/Documentation/adr/0001-multi-platform-packaging.md:
```diff
@@ -0,0 +1,106 @@
+  * `Rx-PlatformServices` becomes `System.Reactive.PlatformServices`
+  * Likewise, the framework-specific libraries `System.Reactive.Windows.Threading`, `System.Reactive.Windows.Forms`, and `System.Reactive.WindowsRuntime` move into eponymous packages
+* v4 made a major change: everything is now in a single `System.Reactive` package
+* v5 stuck with the same structure as v4
```
.NET Framework 4.7.2 was selected as the lowest version for it's .NET Standard 2.0 compatibility and the requirement to be able to use ValueTuple.

idg10 committed:
Update multi-targe ADR

idg10 committed:
A few more tweaks to the ADR

idg10 committed:
Got failing tests passing

* Worked around build failures for UWP runner
* Updated verified APIs to match temporary relocation of WPF and Windows Forms API surface area

Note: although the UWP runner now builds, the tests still don't run

idg10 committed:
Experimentally replace xUnit with MS Test

xUnit doesn't seem to support UWP any more in practice, so this change tries moving everything over to MSTest instead, because it continues to work on VS 2022 with UWP code.

Although this is a fairly mechanical translation (we're not using anything in xUnit's API that doesn't have direct equivalents in other test frameworks), this has uncovered one issue. Unlike xUnit, MSTest doesn't supply a SynchronizationContext, which causes several of the tests to execute difference code paths. Some of them fail as a result.

It's unclear whether this has uncovered latent bugs in Rx, or whether the absence of a synchronization context effectively makes these tests invalid. For now, since this is just an experiment, I've disabled the offending tests.

idg10 committed:
Get UWP project loading by removing net7.0 targets

idg10 committed:
Update multi-platform ADR for post Rx 6.0 release world

idg10 committed:
A few more minor edits to packaging ADR

idg10 committed:
Merge master into experiment/separate-platforms

idg10 committed:
Multiplatform ADR - reworked option 1

idg10 committed:
Merge branch 'main' into experiment/separate-platforms

idg10 committed:
Fix packaging ADR numbering

This was on 0001 because when we started the draft it was the only ADR.
But others were finished first.

idg10 committed:
More detail on design options
